File: buildSrc/src/main/java/net/impactdev/gts/gradle/enums/ReleaseLevel.java
/* This design originates from Nucleus, which is licensed under MIT */
package net.impactdev.gts.gradle.enums;

import java.util.function.Predicate;

public enum ReleaseLevel {

    SNAPSHOT(version -> version.endsWith("SNAPSHOT"), "snapshot", true),
    RELEASE_CANDIDATE(version -> version.contains("RC"), "release-candidate", false),
    MAJOR(version -> version.endsWith(".0.0"), "major", false),
    MINOR(version -> version.endsWith(".0"), "minor", false),
    PATCH(version -> true, "patch", false);

    private final Predicate<String> validator;
    private final String template;
    private final boolean snapshot;

    ReleaseLevel(Predicate<String> validator, String template, boolean snapshot) {
        this.validator = validator;
        this.template = template;
        this.snapshot = snapshot;
    }

    public Predicate<String> getValidator() {
        return this.validator;
    }

    public String getTemplate() {
        return this.template;
    }

    public boolean isSnapshot() {
        return this.snapshot;
    }

    public static ReleaseLevel get(String version) {
        for(ReleaseLevel level : ReleaseLevel.values()) {
            if(level.getValidator().test(version)) {
                return level;
            }
        }

        return ReleaseLevel.SNAPSHOT;
    }

}

File: buildSrc/src/main/kotlin/net/impactdev/gts/gradle/enums/ReleaseLevel.kt
/* This design originates from Nucleus, which is licensed under MIT */
package net.impactdev.gts.gradle.enums

import java.util.function.Predicate

enum class ReleaseLevel(val validator: Predicate<String>, val template: String, val isSnapshot: Boolean) {
    SNAPSHOT(Predicate { version: String -> version.endsWith("SNAPSHOT") }, "snapshot", true), RELEASE_CANDIDATE(
        Predicate { version: String -> version.contains("RC") }, "release-candidate", false
    ),
    MAJOR(
        Predicate { version: String -> version.endsWith(".0.0") }, "major", false
    ),
    MINOR(
        Predicate { version: String -> version.endsWith(".0") }, "minor", false
    ),
    PATCH(
        Predicate<String> { version: String? -> true }, "patch", false
    );

    companion object {
        operator fun get(version: String): ReleaseLevel {
            for (level in values()) {
                if (level.validator.test(version)) {
                    return level
                }
            }
            return SNAPSHOT
        }
    }
}